File: python_email_server/storage.py
from typing import Dict, List, Optional, Any
from datetime import datetime

from .models import EmailAccount, EmailMessage


class Storage:
    """In-memory storage for the email server."""
    
    def __init__(self):
        self.email_accounts: Dict[int, EmailAccount] = {}
        self.email_messages: Dict[int, EmailMessage] = {}
        self.account_current_id = 1
        self.message_current_id = 1
        
        # Create some initial accounts for demo purposes
        self._seed_data()
        
    def _seed_data(self):
        """Add some example email accounts."""
        accounts = [
            {
                "username": "john.doe",
                "domain": "openmail.org",
                "email": "john.doe@openmail.org",
                "password": "password123"
            },
            {
                "username": "dev",
                "domain": "openmail.org",
                "email": "dev@openmail.org",
                "password": "password123"
            },
            {
                "username": "support",
                "domain": "openmail.org",
                "email": "support@openmail.org",
                "password": "password123"
            }
        ]
        
        for account_data in accounts:
            self.create_email_account(account_data)
            
    # Email Account Methods
    def get_email_accounts(self) -> List[EmailAccount]:
        """Get all email accounts."""
        return list(self.email_accounts.values())
    
    def get_email_account(self, account_id: int) -> Optional[EmailAccount]:
        """Get an email account by ID."""
        return self.email_accounts.get(account_id)
    
    def get_email_account_by_email(self, email: str) -> Optional[EmailAccount]:
        """Get an email account by email address."""
        email = email.lower()
        for account in self.email_accounts.values():
            if account.email.lower() == email:
                return account
        return None
    
    def create_email_account(self, account_data: Dict[str, Any]) -> EmailAccount:
        """Create a new email account."""
        account_id = self.account_current_id
        self.account_current_id += 1
        
        account = EmailAccount(
            id=account_id,
            username=account_data["username"],
            domain=account_data["domain"],
            email=account_data["email"],
            password=account_data["password"],
            created_at=datetime.now()
        )
        
        self.email_accounts[account_id] = account
        return account
    
    # Email Message Methods
    def get_email_messages(self, account_id: int) -> List[EmailMessage]:
        """Get all email messages for an account."""
        messages = [
            msg for msg in self.email_messages.values()
            if msg.account_id == account_id
        ]
        # Sort by received_at, newest first
        return sorted(messages, key=lambda x: x.received_at, reverse=True)
    
    def get_email_message(self, message_id: int) -> Optional[EmailMessage]:
        """Get an email message by ID."""
        return self.email_messages.get(message_id)
    
    def create_email_message(self, message_data: Dict[str, Any]) -> EmailMessage:
        """Create a new email message."""
        message_id = self.message_current_id
        self.message_current_id += 1
        
        # Ensure html_content is string or None
        html_content = message_data.get("html_content")
        if html_content is None and "html_content" in message_data:
            html_content = None
            
        # Ensure headers is a dict
        headers = message_data.get("headers", {})
        
        message = EmailMessage(
            id=message_id,
            account_id=message_data["account_id"],
            sender=message_data["sender"],
            sender_email=message_data["sender_email"],
            recipient=message_data["recipient"],
            subject=message_data["subject"],
            content=message_data["content"],
            html_content=html_content,
            headers=headers,
            received_at=datetime.now(),
            read=False
        )
        
        self.email_messages[message_id] = message
        return message
    
    def mark_email_as_read(self, message_id: int) -> EmailMessage:
        """Mark an email message as read."""
        message = self.get_email_message(message_id)
        if not message:
            raise ValueError(f"Email message with id {message_id} not found")
        
        message.read = True
        return message
    
    def get_unread_count(self, account_id: int) -> int:
        """Get the number of unread email messages for an account."""
        return sum(
            1 for msg in self.email_messages.values()
            if msg.account_id == account_id and not msg.read
        )


# Create a singleton instance of Storage
storage = Storage()

File: server/storage.ts
import { 
  EmailAccount, 
  InsertEmailAccount, 
  EmailMessage, 
  InsertEmailMessage 
} from "@shared/schema";

export interface IStorage {
  // Email account operations
  getEmailAccounts(): Promise<EmailAccount[]>;
  getEmailAccount(id: number): Promise<EmailAccount | undefined>;
  getEmailAccountByEmail(email: string): Promise<EmailAccount | undefined>;
  createEmailAccount(account: InsertEmailAccount): Promise<EmailAccount>;
  
  // Email message operations
  getEmailMessages(accountId: number): Promise<EmailMessage[]>;
  getEmailMessage(id: number): Promise<EmailMessage | undefined>;
  createEmailMessage(message: InsertEmailMessage): Promise<EmailMessage>;
  markEmailAsRead(id: number): Promise<EmailMessage>;
  getUnreadCount(accountId: number): Promise<number>;
}

export class MemStorage implements IStorage {
  private emailAccounts: Map<number, EmailAccount>;
  private emailMessages: Map<number, EmailMessage>;
  private accountCurrentId: number;
  private messageCurrentId: number;

  constructor() {
    this.emailAccounts = new Map();
    this.emailMessages = new Map();
    this.accountCurrentId = 1;
    this.messageCurrentId = 1;
    
    // Create some initial accounts for demo purposes
    this.seedData();
  }

  private seedData() {
    // Add some example email accounts
    const accounts = [
      {
        username: "john.doe",
        domain: "openmail.org",
        email: "john.doe@openmail.org",
        password: "password123"
      },
      {
        username: "dev",
        domain: "openmail.org",
        email: "dev@openmail.org",
        password: "password123"
      },
      {
        username: "support",
        domain: "openmail.org",
        email: "support@openmail.org",
        password: "password123"
      }
    ];
    
    accounts.forEach(account => this.createEmailAccount(account));
  }

  // Email Account Methods
  async getEmailAccounts(): Promise<EmailAccount[]> {
    return Array.from(this.emailAccounts.values());
  }

  async getEmailAccount(id: number): Promise<EmailAccount | undefined> {
    return this.emailAccounts.get(id);
  }

  async getEmailAccountByEmail(email: string): Promise<EmailAccount | undefined> {
    return Array.from(this.emailAccounts.values()).find(
      account => account.email.toLowerCase() === email.toLowerCase()
    );
  }

  async createEmailAccount(insertAccount: InsertEmailAccount): Promise<EmailAccount> {
    const id = this.accountCurrentId++;
    const createdAt = new Date();
    const account: EmailAccount = { ...insertAccount, id, createdAt };
    this.emailAccounts.set(id, account);
    return account;
  }

  // Email Message Methods
  async getEmailMessages(accountId: number): Promise<EmailMessage[]> {
    return Array.from(this.emailMessages.values())
      .filter(message => message.accountId === accountId)
      .sort((a, b) => b.receivedAt.getTime() - a.receivedAt.getTime()); // Newest first
  }

  async getEmailMessage(id: number): Promise<EmailMessage | undefined> {
    return this.emailMessages.get(id);
  }

  async createEmailMessage(insertMessage: InsertEmailMessage): Promise<EmailMessage> {
    const id = this.messageCurrentId++;
    const receivedAt = new Date();
    const read = false;
    // Ensure htmlContent is always string | null
    const htmlContent = insertMessage.htmlContent === undefined ? null : insertMessage.htmlContent;
    // Ensure headers is not undefined
    const headers = insertMessage.headers || {};
    
    const message: EmailMessage = { 
      ...insertMessage, 
      id, 
      receivedAt, 
      read,
      htmlContent,
      headers
    };
    
    this.emailMessages.set(id, message);
    return message;
  }

  async markEmailAsRead(id: number): Promise<EmailMessage> {
    const message = this.emailMessages.get(id);
    if (!message) {
      throw new Error(`Email message with id ${id} not found`);
    }
    
    const updatedMessage = { ...message, read: true };
    this.emailMessages.set(id, updatedMessage);
    return updatedMessage;
  }

  async getUnreadCount(accountId: number): Promise<number> {
    return Array.from(this.emailMessages.values())
      .filter(message => message.accountId === accountId && !message.read)
      .length;
  }
}

export const storage = new MemStorage();
